apps/user-ms/src/main/java/it/pagopa/selfcare/user/service/UserServiceImpl.java

```diff
@@ -131,4 +131,32 @@ public Uni<Void> deleteUserInstitutionProduct(String userId, String institutionI
                 });
     }
 
+    @Override
+    public Uni<List<UserInstitution>> retrieveBindings(String institutionId, String userId, String[] states) {
+        String[] finalStates = states != null && states.length > 0 ? states : null;
+        List<OnboardedProductState> relationshipStates = Optional.ofNullable(finalStates)
+                .map(userUtils::convertStatesToRelationshipsState)
+                .orElse(null);
+
+        UserInstitutionFilter userInstitutionFilter = UserInstitutionFilter.builder()
+                .userId(userId)
+                .institutionId(institutionId)
+                .build();
+
+        OnboardedProductFilter onboardedProductFilter = OnboardedProductFilter.builder()
+                .status(relationshipStates)
+                .build();
+
+        return userInstitutionService.retrieveFilteredUserInstitution(userUtils.retrieveMapForFilter(userInstitutionFilter.constructMap(), onboardedProductFilter.constructMap()))
+                .map((list) -> {
+                    if(list == null || list.isEmpty()) {
+                        throw new ResourceNotFoundException("");
+                    }
+                    return list;
+                })
+                .onItem().transformToMulti(Multi.createFrom()::iterable)
+                .map(userInstitution -> userUtils.filterProduct(userInstitution, finalStates))
+                .collect()
+                .asList();
```

Comment:
Do you use "finalStates" filter twice? I look at retrieveFilteredUserInstitution above `retrieveFilteredUserInstitution(..., onboardedProductFilter.constructMap()))`

Reply:
Yes I use it twice, because in the first case I need it to get from the query all UserInstitutions in which there are products with the status passed in input. But the query returns a UserInstitution with all products, even though a status filter was given that should exclude some of them. So in the second case, the "filterProduct" method removes from the UserInstitution object all products that do not have the status given in the filter.

Reply:
ok you could add part of this comment at line 157....

In addition, you could avoid to transform in multi and then to uni using stream method. For example at line 157:

onItem().transform(userInstitutionList -> userInstuserInstitutionListitutions.stream().map(userInstitution -> userUtils.filterProduct(userInstitution, finalStates)).collect(Collections.toList()))